tripal/src/Entity/TripalEntity.php

```diff
@@ -174,36 +194,61 @@ public function getDefaultAlias() {
     return $default_alias;
   }
 
+  /**
+   * Returns the URL alias for the current entity.
+   *
+   * @return string
+   *   The URL alias e.g. "/organism/123"
+   */
+  public function getAlias() {
+    $system_path = '/bio_data/' . $this->getID();
+    $langcode = $this->defaultLangcode;
+    $existing_alias = \Drupal::service('path_alias.repository')->lookupBySystemPath($system_path, $langcode);
+    return $existing_alias;
+  }
+
   /**
    * Sets a URL alias for the current entity if one does not already exist.
    *
    * @param string $path_alias
    *   The alias to use. It can contain tokens that correspond to field values.
    *   Tokens should be be compatible with those returned by
-   *   tripal_get_entity_tokens(). If NULL, then use the default alias.
+   *   tripal_get_entity_tokens(). If empty, then use the default alias template.
+   *   If $path_alias is specified, then any existing alias will be updated.
+   *
+   * @return string
+   *   Returns the path alias that was used with tokens replaced
    */
-  public function setDefaultAlias($path_alias = NULL) {
+  public function setAlias(string $path_alias = ''): string {
+    // Check if an alias already exists for this entity's system path
+    $existing_alias = $this->getAlias();
 
-    if (!$path_alias) {
-      $path_alias = $this->getDefaultAlias();
-    }
-    $system_path = "/bio_data/" . $this->getID();
-    $langcode = $this->defaultLangcode;
+    // @todo check if  alias exists for somthing else?
 
-    // Check if an alias already exists.
-    $alias_exists = FALSE;
-    if (\Drupal::service('path_alias.repository')->lookupBySystemPath($system_path, $langcode)) {
-      $alias_exists = TRUE;
-    }
+    // Gets and uses default template, or replaces tokens
+    // in the supplied $path_alias.
+    $new_alias = $this->getDefaultAlias($path_alias);
+
+    // If alias does not exist, or if it is being changed
+    if (!$existing_alias or ($existing_alias['alias'] != $new_alias)) {
 
-    // If an alias does not exist, then create a default alias.
-    if (!$alias_exists) {
-      $path = \Drupal::entityTypeManager()->getStorage('path_alias')->create([
+      // If an alias already exists, we first need to remove it
+      if ($existing_alias) {
+        $alias_objects = \Drupal::entityTypeManager()->getStorage('path_alias')->loadByProperties([
+          'alias' => $existing_alias['alias'],
+        ]);
+        $alias_objects[array_key_first($alias_objects)]->delete();
```

Comment:
We need more error checking on this line in case something goes wrong along the way. For example, you should check that $alias_objects is a non-empty array before grabbing the first key and you should check that `$alias_objects[array_key_first($alias_objects)]` is actually an object of the correct type. If either of these are not true you should throw an exception and this exception should be caught by the post save in order to provide a useful error logged to the user.

Reply:
This part changed significantly with commit https://github.com/tripal/tripal/pull/2040/commits/30d93853cc7122d1384aa90f8ec6f75c305ab599 - instead of fetching the alias, deleting, and creating a new one, I now (as I saw in Drupal core) just update the existing alias. But I have some error checking in there now